pyccl/nl_pt/lpt.py

```diff
@@ -0,0 +1,492 @@
+
+        # a sampling
+        if a_arr is None:
+            a_arr = get_pk_spline_a()
+        self.a_s = a_arr.copy()
+        self.z_s = 1/self.a_s-1
+
+        # Cutoff factor
+        if k_cutoff is not None:
+            self.exp_cutoff = np.exp(-(self.k_s/k_cutoff)**n_exp_cutoff)
+            self.exp_cutoff = self.exp_cutoff[None, :]
+        else:
+            self.exp_cutoff = 1
+
+        # b1/bk P(k) prescription
+        if b1_pk_kind not in ['linear', 'nonlinear', 'pt']:
+            raise ValueError(f"Unknown P(k) prescription {b1_pk_kind}")
+        if bk2_pk_kind not in ['linear', 'nonlinear', 'pt']:
+            raise ValueError(f"Unknown P(k) prescription {bk2_pk_kind}")
+        self.b1_pk_kind = b1_pk_kind
+        self.bk2_pk_kind = bk2_pk_kind
+
+        # Initialize all expensive arrays to ``None``.
+        self._cosmo = None
+
+        # Fill them out if cosmo is present
+        if cosmo is not None:
+            self.update_ingredients(cosmo)
+
+        # All valid Pk pair labels
+        self._pk_valid = list(_PK_ALIAS.keys())
+        # List of Pk2Ds to fill out
+        self._pk2d_temp = {}
+
+    def _check_init(self):
+        if self.initialised:
+            return
+        raise CCLError("PT templates have not been initialised "
+                       "for this calculator. Please do so using "
+                       "`update_ingredients`.")
+
+    @property
+    def initialised(self):
+        return hasattr(self, "pk_bk")
+
+    @unlock_instance
+    def update_ingredients(self, cosmo):
+        """ Update the internal PT arrays.
+
+        Args:
+            cosmo (:class:`~pyccl.cosmology.Cosmology`): a Cosmology object.
+        """
+        if self.initialised and (cosmo == self._cosmo):
+            return
+
+        pklz0 = cosmo.linear_matter_power(self.k_s, 1.0)
+        g = cosmo.growth_factor(self.a_s)
+
+        from velocileptors.EPT.cleft_kexpanded_resummed_fftw import RKECLEFT
+        h = cosmo['h']
+        cleft = RKECLEFT(self.k_s / h, pklz0 * h ** 3)
+        lpt_table = []
+        for gz in g:
+            cleft.make_ptable(D=gz, kmin=self.k_s[0] / h,
+                              kmax=self.k_s[-1] / h, nk=self.k_s.size)
+            lpt_table.append(cleft.pktable)
+        lpt_table = np.array(lpt_table)
+        lpt_table[:, :, 1:] /= h ** 3
+
+        self.lpt_table = lpt_table
+        self.one_loop_dd = self.lpt_table[:, :, 1]
+
+        # b1/bk power spectrum
+        pks = {}
+        if 'nonlinear' in [self.b1_pk_kind, self.bk2_pk_kind]:
+            pks['nonlinear'] = np.array([cosmo.nonlin_matter_power(self.k_s, a)
+                                         for a in self.a_s])
+        if 'linear' in [self.b1_pk_kind, self.bk2_pk_kind]:
+            pks['linear'] = np.array([cosmo.linear_matter_power(self.k_s, a)
+                                      for a in self.a_s])
+        if 'pt' in [self.b1_pk_kind, self.bk2_pk_kind]:
+            pks['pt'] = None
```

Comment:
can you clarify what this implies in a comment?

Reply:
Done.